buildpacks/nodejs-yarn/src/main.rs

```diff
@@ -0,0 +1,232 @@
+        log_header("Installing dependencies");
+        cmd::yarn_install(&yarn, zero_install, &env).map_err(YarnBuildpackError::YarnInstall)?;
+
+        log_header("Running scripts");
+        let scripts = cfg::get_build_scripts(&pkg_json);
+        if scripts.is_empty() {
+            log_info("No build scripts found");
+        } else {
+            for script in scripts {
+                log_info(format!("Running `{script}` script"));
+                cmd::yarn_run(&env, &script).map_err(YarnBuildpackError::BuildScript)?;
+            }
+        }
+
+        if cfg::has_start_script(&pkg_json) {
+            BuildResultBuilder::new()
+                .launch(
+                    LaunchBuilder::new()
+                        .process(
+                            ProcessBuilder::new(process_type!("web"), "yarn")
+                                .arg("start")
+                                .default(true)
+                                .build(),
+                        )
+                        .build(),
+                )
+                .build()
+        } else {
+            BuildResultBuilder::new().build()
+        }
+    }
+
+    fn on_error(&self, error: libcnb::Error<Self::Error>) {
+        match error {
+            libcnb::Error::BuildpackError(bp_err) => {
+                let err_string = bp_err.to_string();
+                match bp_err {
+                    YarnBuildpackError::BuildScript(_) => {
+                        log_error("Yarn build script error", err_string);
+                    }
+                    YarnBuildpackError::CliLayer(_) => {
+                        log_error("Yarn distribution layer error", err_string);
+                    }
+                    YarnBuildpackError::DepsLayer(_) => {
+                        log_error("Yarn dependency layer error", err_string);
+                    }
+                    YarnBuildpackError::InventoryParse(_) => {
+                        log_error("Yarn inventory parse error", err_string);
+                    }
+                    YarnBuildpackError::PackageJson(_) => {
+                        log_error("Yarn package.json error", err_string);
+                    }
+                    YarnBuildpackError::YarnCacheSet(_) | YarnBuildpackError::YarnCacheGet(_) => {
+                        log_error("Yarn cache error", err_string);
+                    }
+                    YarnBuildpackError::YarnInstall(_) => {
+                        log_error("Yarn install error", err_string);
+                    }
+                    YarnBuildpackError::YarnVersionDetect(_)
+                    | YarnBuildpackError::YarnVersionResolve(_)
+                    | YarnBuildpackError::YarnVersionUnsupported(_) => {
+                        log_error("Yarn version error", err_string);
+                    }
+                }
+            }
+            err => {
+                log_error("Yarn internal buildpack error", err.to_string());
+            }
+        }
+    }
```

Comment:
Have you considered decoupling the error values from the error messages more? I see you're using `thiserror` to derive some basic traits for your error types. This `on_error` implementation uses the `Display` representation for these errors almost exclusively. This means that the actual error messages for our users are distributed across the code-base (`Display` implementation next to the type declarations + prefixes in `on_error`). 

If we would drop `Display` for our errors, we would be forced to put all error messages in one place (maybe even a dedicated module that takes care of mapping errors to user output), isolating this concern from the rest of the buildpack.

I think this approach has several advantages:

- If we ever want to swap out all error messages conditionally we can do that easily since they're all in one module/function. Use cases for this could be different error messages when the buildpack runs on Heroku, locally or for Salesforce Functions. A more unlikely use-case would be translations to a different (human) language.
- It allows us to craft our error messages holistically, both for individual messages but also all of the possible errors in one buildpack. Since we would then see the whole error message at a glance, it's easier to spot formatting, spelling and grammatical issues. When `Display` is used for nested errors, the topmost `Display` might produce strings like "foo bar error: lolwut error: io error: File not found". Explicitly matching against these error trees will yield better error messages.
- Separation of concerns, UI (which these error strings are) would then be strictly separated. We already decouple error logging by returning `Result` instead of having conditionals + `println!` interwoven with the business logic. Moving also the actual strings away from the implementation into one place would bring this concept to conclusion.

If I remember correctly, @edmorley goes with the same style and [the `errors` module for `heroku/maven`](https://github.com/heroku/buildpacks-jvm/blob/1fd4e44152950657e65d94bb007c00f4fc6e09e9/buildpacks/maven/src/errors.rs) uses this approach too. Some of the JVM buildpacks still rely on `Display` for errrors, but eventually I want to get rid of those as well.

This is not a blocker for this PR, we can easily change this later if we want to. Mainly interested about your thoughts on this topic. :)

Reply:
Yes, I tend to agree with your suggestion. The logic for the error messaging is a bit scattered as it is. The error messages aren't as well composed either, with many error's display trait consuming another error's display trait. 

I worry that having to handle every variant, sub-variant, and sub-variant's sub-variant means a huge list of `match`es if we were to refactor. But it's probably worth a shot. I'll tackle this in another PR or maybe try it in the next buildpack.